independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/DecoratedBeanMetadataProvider.java

```diff
@@ -0,0 +1,28 @@
+package io.quarkus.arc.impl;
+
+import static io.quarkus.arc.impl.CreationalContextImpl.unwrap;
+
+import jakarta.enterprise.context.spi.Contextual;
+import jakarta.enterprise.context.spi.CreationalContext;
+import jakarta.enterprise.inject.Decorated;
+import jakarta.enterprise.inject.spi.Bean;
+
+import io.quarkus.arc.InjectableReferenceProvider;
+
+/**
+ * {@link Decorated} {@link Bean} metadata provider.
+ */
+public class DecoratedBeanMetadataProvider implements InjectableReferenceProvider<Contextual<?>> {
```

Comment:
As this is a copy of `InterceptedBeanMetadataProvider`, I'd suggest renaming the original class to something like `InterceptedDecoratedBeanMetadataProvider` and adjusting comments in it to explain it is used for both.
With that, you can also get rid of the extra method `BuiltinBean#generateDecoratedBeanBytecode` and perform similar change for the original method.

Reply:
@manovotn it's done, should I squash everything ?